pilot/pkg/xds/ads.go

```diff
@@ -305,6 +305,27 @@ func (s *DiscoveryServer) Stream(stream DiscoveryStream) error {
 	<-con.initialized
 
 	for {
+		// Go select{} statements are not ordered; the same channel can be chosen many times.
+		// For requests, these are higher priority (client may be blocked on startup until these are done)
+		// and often very cheap to handle (simple ACK), so we check it first.
+		select {
+		case req, ok := <-con.reqChan:
+			if ok {
+				if err := s.processRequest(req, con); err != nil {
+					return err
+				}
+			} else {
+				// Remote side closed connection or error processing the request.
+				return <-con.errorChan
+			}
```

Comment:
need to break in this case, and do another recvChan read

Reply:
for which part? the !ok?

Reply:
if ok{
goto L311 again
}

Reply:
`goto L311 again` can be put at end of this case (case req, ok := <-con.reqChan:)

Reply:
This is not done intentionally, see comment below:
```
		// If there wasn't already a request, poll for requests and pushes. Note: if we have a huge
		// amount of incoming requests, we may still send some pushes, as we do not `continue` above;
		// however, requests will be handled ~2x as much as pushes. This ensures a wave of requests
		// cannot completely starve pushes. However, this scenario is unlikely.
```

Reply:
I understand your meaning, this is a tradeoff, hard to say which is better.

Say if request is not handled, we donot have a chance to update the watched resources, then the push handling could be meaningless

Reply:
Yeah definitely it is a tradeoff. On the other hand, we may have requests for new resource (applying some new config, doesn't need to be super fast) blocking an EDS push (critical).

I am not sure there is a perfect solution. I went back and forth on this but went with the current approach because its closer to the old way (fully random) so seemed less risky

Reply:
The overall experiment is excellent, great work.

I kind of remember priority channel somewhere, could not find it.